crates/red_knot_project/src/metadata/pyproject.rs

```diff
@@ -55,6 +56,75 @@ pub struct Project {
     pub requires_python: Option<RangedValue<VersionSpecifiers>>,
 }
 
+impl Project {
+    pub(super) fn resolve_requires_python_lower_bound(
+        &self,
+    ) -> Result<Option<RangedValue<PythonVersion>>, ResolveRequiresPythonError> {
+        let Some(requires_python) = self.requires_python.as_ref() else {
+            return Ok(None);
+        };
+
+        tracing::debug!("Resolving requires-python constraint: `{requires_python}`");
+
+        let ranges = release_specifiers_to_ranges((**requires_python).clone());
+        let Some((lower, _)) = ranges.bounding_range() else {
+            return Ok(None);
+        };
+
+        let version = match lower {
+            // Ex) `>=3.10.1` -> `>=3.10`
+            Bound::Included(version) => version,
+
+            // Ex) `>3.10.1` -> `>=3.10` or `>3.10` -> `>=3.10`
+            // The second example looks obscure at first but it is required because
+            // `3.10.1 > 3.10` is true but we only have two digits here. So including 3.10 is the
+            // right move. Overall, using `>` without a patch release is most likely bogus.
+            Bound::Excluded(version) => version,
+
+            // Ex) `<3.10` or ``
+            Bound::Unbounded => {
+                return Err(ResolveRequiresPythonError::NoLowerBound(
+                    requires_python.to_string(),
+                ))
+            }
+        };
+
+        // Take the major and minor version
+        let mut versions = version.release().iter().take(2);
+
+        let Some(major) = versions.next().copied() else {
+            return Ok(None);
+        };
+
+        let minor = versions.next().copied().unwrap_or_default();
+
+        tracing::debug!("Resolved requires-python constraint to: {major}.{minor}");
+
+        let major =
+            u8::try_from(major).map_err(|_| ResolveRequiresPythonError::TooLargeMajor(major))?;
+        let minor =
+            u8::try_from(minor).map_err(|_| ResolveRequiresPythonError::TooLargeMajor(minor))?;
+
+        let python_version = PythonVersion::from((major, minor));
+
+        Ok(Some(if let Some(range) = requires_python.range() {
+            RangedValue::with_range(python_version, requires_python.source().clone(), range)
+        } else {
+            RangedValue::new(python_version, requires_python.source().clone())
+        }))
```

Comment:
nit: I'd find it a bit easier to see what's different between the branches if we moved `source` out

```suggestion
        let python_version = PythonVersion::from((major, minor));
        let source = requires_python.source().clone();

        Ok(Some(if let Some(range) = requires_python.range() {
            RangedValue::with_range(python_version, source, range)
        } else {
            RangedValue::new(python_version, source)
        }))
```